lib/policy_machine_storage_adapters/active_record/postgresql.rb

```diff
@@ -2,6 +2,42 @@
 
 module PolicyMachineStorageAdapter
   class ActiveRecord
+    class PolicyElementAssociation
+      def self.with_accessible_operation(associations, operation)
+        query = <<-SQL
+          operation_set_id IN (
+            WITH RECURSIVE accessible_operations AS (
+              (
+                SELECT
+                  child_id,
+                  parent_id,
+                  parent_id AS operation_set_id
+                FROM assignments
+                WHERE parent_id IN (#{associations.select(:operation_set_id).to_sql})
+              )
+              UNION ALL
+              (
+                SELECT
+                  assignments.child_id,
+                  assignments.parent_id,
+                  accessible_operations.operation_set_id AS operation_set_id
+                FROM assignments
+                INNER JOIN accessible_operations
+                ON accessible_operations.child_id = assignments.parent_id
+              )
+            )
+          SELECT accessible_operations.operation_set_id
+          FROM accessible_operations
+          JOIN policy_elements ops
+            ON ops.id = accessible_operations.child_id
+          WHERE ops.unique_identifier = ?
+          )
+        SQL
+
+        associations.where(query, operation)
```

Comment:
also maybe it is faster to join directly to the result temp table from the recursive query output instead of us it as a subquery? I'm not sure I remember ever trying that though I've made recursive temp tables 🤔

Reply:
That's an excellent idea, I would be very surprised if that wasn't more efficient. I'll keep that in mind when I revisit this.

Reply:
if you do an EXPLAIN_ANALYZE you'll notice that materializing subqueries of arrays of IDs is weirdly slow